core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -429,111 +439,70 @@ private boolean deassociateDeadNodes(RoutingAllocation allocation) {
+                routingNodes.remove(relocationSourceShard);
             }
         }
-        return dirty;
     }
 
     /**
-     * Applies the relevant logic to handle a failed shard. Returns <tt>true</tt> if changes happened that
-     * require relocation.
+     * Applies the relevant logic to handle a failed shard.
      */
-    private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting failedShard, boolean addToIgnoreList, UnassignedInfo unassignedInfo) {
-        IndexRoutingTable indexRoutingTable = allocation.routingTable().index(failedShard.index());
-        if (indexRoutingTable == null) {
-            logger.debug("{} ignoring shard failure, unknown index in {} ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
-            return false;
-        }
+    private void applyFailedShard(RoutingAllocation allocation, ShardRouting failedShard, UnassignedInfo unassignedInfo) {
         RoutingNodes routingNodes = allocation.routingNodes();
-
-        RoutingNode matchedNode = routingNodes.node(failedShard.currentNodeId());
-        if (matchedNode == null) {
-            logger.debug("{} ignoring shard failure, unknown node in {} ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
-            return false;
-        }
-
-        ShardRouting matchedShard = matchedNode.getByShardId(failedShard.shardId());
-        if (matchedShard != null && matchedShard.isSameAllocation(failedShard)) {
-            logger.debug("{} failed shard {} found in routingNodes, failing it ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
-            // replace incoming instance to make sure we work on the latest one
-            failedShard = matchedShard;
-        } else {
-            logger.debug("{} ignoring shard failure, unknown allocation id in {} ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
-            return false;
-        }
-
+        assert failedShard.assignedToNode() : "only assigned shards can be failed";
+        assert allocation.metaData().index(failedShard.shardId().getIndex()) != null :
+            "shard failed for unknown index (shard entry: " + failedShard + ")";
+        assert routingNodes.getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId()) == failedShard :
+            "shard routing to fail does not exist in routing table, expected: " + failedShard + " but was: " +
+                routingNodes.getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId());
+
+        logger.debug("{} failing shard {} with unassigned info ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
         if (failedShard.primary()) {
             // fail replicas first otherwise we move RoutingNodes into an inconsistent state
             failReplicasForUnassignedPrimary(allocation, failedShard);
         }
 
-        if (addToIgnoreList) {
```

Comment:
it was not removed. I moved it to the `applyFailedShards` method (https://github.com/elastic/elasticsearch/pull/19715/files#diff-792ac9952eb61524ecac9d045d6eae4bR248). That method was the only one calling `applyFailedShard` with `addToIgnoreList = true`.